File: app/src/main/java/com/dsy/mvp/base/BaseFragment.java
package com.dsy.mvp.base;

import android.app.Activity;
import android.app.ProgressDialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.blankj.utilcode.util.ToastUtils;
import com.dsy.mvp.base.impl.IPresenter;
import com.dsy.mvp.base.impl.IView;

import butterknife.ButterKnife;
import butterknife.Unbinder;

public abstract class BaseFragment<T extends IPresenter> extends Fragment implements IView {
    protected View view;
    protected Bundle savedInstanceState;
    protected T mPresenter;
    protected String TAG = this.getClass().getSimpleName();
    private Unbinder unbinder;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        mPresenter = createPresenter();
        attachView();
        super.onCreate(savedInstanceState);
    }


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        this.savedInstanceState = savedInstanceState;
        if (null == view) {
            view = inflater.inflate(createLayoutId(), container, false);
            unbinder = ButterKnife.bind(this, view);
            initView();
            initEvent();
            initData();
        }
        ViewGroup parent = (ViewGroup) view.getParent();
        if (parent != null) {
            parent.removeView(view);
        }
        return view;
    }

    /**
     * @return LayoutId
     */
    public abstract int createLayoutId();

    /**
     * P层绑定   若无则返回null;
     */
    protected abstract T createPresenter();

    /**
     * 控件绑定
     */
    protected abstract void initView();

    /**
     * 事件触发绑定
     */
    protected abstract void initEvent();

    /**
     * 数据初始化
     */
    protected abstract void initData();

    /**
     * P层绑定V层
     */
    private void attachView() {
        if (null != mPresenter) {
            mPresenter.attachView(this);
        }
    }

    /**
     * P层解绑V层
     */
    private void detachView() {
        if (null != mPresenter) {
            mPresenter.detachView();
        }
    }

    @Override
    public Activity getMActivity() {
        return getActivity();
    }

    @Override
    public void toast(String msg) {
        ToastUtils.showShort(msg);
    }

    @Override
    public void toast(int id) {
        ToastUtils.showShort(id);
    }


    public ProgressDialog progressDialog;

    @Override
    public void showLoadingDialog() {
        progressDialog = new ProgressDialog(getActivity());
        progressDialog.setMessage("加载中");
        progressDialog.show();
    }

    @Override
    public void hideLoadingDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        detachView();
        unbinder.unbind();
    }


}

File: app/src/main/java/com/dsy/mvp/base/BaseFragment.kt
package com.dsy.mvp.base

import android.app.Activity
import android.app.ProgressDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import com.blankj.utilcode.util.ToastUtils
import com.dsy.mvp.base.impl.IPresenter
import com.dsy.mvp.base.impl.IView
import com.dsy.mvp.widget.status.LoadStatus
import com.dsy.mvp.widget.status.StatusHolder
import com.dylanc.loadinghelper.LoadingHelper
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

abstract class BaseFragment<T : IPresenter>(private val layoutId: Int) : Fragment(), IView {
    private var rootView: View? = null
    protected lateinit var mPresenter: T
    protected var TAG = this.javaClass.simpleName
    protected var isViewBind = false
    private var isFirstVisible = true //第一次展示
    protected lateinit var mLoadingHelper:LoadingHelper
    protected lateinit var mStatusHolder: StatusHolder.Holder
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenter = initPresenter()
        mPresenter.attachView(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (null == rootView) {
            val view = inflater.inflate(layoutId, container, false)
            if (getStatusWrapView(view) == null){
                mStatusHolder = StatusHolder.getDefault().wrap(view)
                rootView = mStatusHolder.wrapper
            }else{
                mStatusHolder = StatusHolder.getDefault().wrap(getStatusWrapView(view))
                rootView = view
            }
            mStatusHolder.withRetry { onReload() }
            isViewBind = false
        }
//        val parent = rootView?.parent as ViewGroup?
//        parent?.removeView(rootView)
        return mLoadingHelper.decorView
    }

    /**
     * 状态布局包裹的view
     */
    protected open fun getStatusWrapView(rootView: View): View? {
        return null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (!isViewBind) {
            initView()
            initEvent()
            isViewBind = true
        }
    }

    @Suppress("UNCHECKED_CAST")
    protected fun initPresenter(): T {
        val type: Type? = javaClass.genericSuperclass
        val modelClass = if (type is ParameterizedType) {
            type.actualTypeArguments[0] as Class<*>
        } else {
            //如果没有指定泛型参数，则默认使用EmptyPresenter
            SimplePresenter::class.java
        }
        return modelClass.newInstance() as T
    }

    override fun showLoadingView() {
        mStatusHolder.showLoading()
    }

    override fun showContentView() {
        mStatusHolder.showLoadSuccess()
    }

    override fun showErrorView() {
        mStatusHolder.showLoadFailed()
    }

    override fun showEmptyView() {
        mStatusHolder.showEmpty()
    }

    override fun showCustomView(status: LoadStatus) {
        mStatusHolder.showLoadingStatus(status)
    }


    /**
     * 控件绑定
     */
    protected abstract fun initView()

    /**
     * 事件触发绑定
     */
    protected abstract fun initEvent()

    /**
     * 数据初始化
     */
    protected abstract fun initData()

    protected open fun onReload() {}

    override val mActivity: Activity
        get() = activity!!

    override fun toast(msg: CharSequence) {
        ToastUtils.showShort(msg)
    }

    override fun toast(@StringRes id: Int) {
        ToastUtils.showShort(id)
    }

    private var progressDialog: ProgressDialog? = null
    override fun showLoadingDialog() {
        progressDialog = ProgressDialog(activity)
        progressDialog?.setMessage("加载中")
        progressDialog?.show()
    }

    override fun hideLoadingDialog() {
        progressDialog?.dismiss()
    }

    override fun onDestroy() {
        mPresenter.detachView()
        super.onDestroy()
    }
}